runners/core-java/src/main/java/org/apache/beam/runners/core/SplittableParDo.java

```diff
@@ -138,6 +136,15 @@ private PCollectionTuple applyTyped(PCollection<InputT> input) {
             .setCoder(splitCoder)
             .apply("Split restriction", ParDo.of(new SplitRestrictionFn<InputT, RestrictionT>(fn)))
             .setCoder(splitCoder)
+            // We need to explode windows before the GBK, so that the elements in KeyedWorkItem's
+            // are each in a single window. This is necessary because ProcessFn sets timers, and
+            // timers are namespaced to a single window and it should be the window of the input
+            // element. See also: https://issues.apache.org/jira/browse/BEAM-1983
```

Comment:
Most of this probably can move to the documentation of `ProcessFn` - specifically, note here
"`ProcessFn` requires all input elements to be in a single window and have a single element per work item. This must precede the unique keying so each key has a single associated element"